### ISSUE TITLE
tsconfig-paths rewrite to be broken when import with file extension (.js)

### ISSUE THREAD

**Issue by msimon** [nestjs/nest-cli#1437]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current behavior

With a project organised that way:
```
src:
  - main.ts
  - app.module.ts
  ....
tsconfig.json
...
```

Defining an import in `main.ts` the following way:
```
import { AppModule } from 'src/app.module.js';
```
result in it being compiled to:
```
const app_module_1 = require("src/app.module.js");
```
ignoring the non-relative path to relative path rewrite that should be done by `tsconfig-paths`.

Removing the `.js` from the import, result in the correct compilation:
```
import { AppModule } from 'src/app.module';
```
```
const app_module_1 = require("./app.module");
```

I took a quick look into `nest-cli/lib/compiler/hooks/tsconfig-paths.hook.ts` to see if I could spot the issues.
My guess is that the match on l.56 does not return anything if the extension is set.

I'll try to ding more into it soon, but I figure I would ask in case there is an obvious fix for this.

### Minimum reproduction code

.

### Steps to reproduce

After adding/removing the .js
```
nest build --tsc
cat dist/main.js
```

### Expected behavior

Defining an import in `main.ts` the following way:
```
import { AppModule } from 'src/app.module.js';
```
Should compile to:
```
const app_module_1 = require("./app.module.js");
```

### Package version

8.1.5

### NestJS version

8.2.3

### Node.js version

v14.17.0

### In which operating systems have you tested?

- [X] macOS
- [ ] Windows
- [ ] Linux

**Comment by jmcdo29** [comment]
Can you please provide a minimum reproduction repository?

**Comment by kamilmysliwiec** [comment]
BTW using absolute paths (using "src" in your import paths) is considered as a bad practice

**Comment by msimon** [comment]
@jmcdo29 https://github.com/HomeskilletHealthInc/nestJs_import_rewrite_bug

@kamilmysliwiec Using absolute path is, but I don't believe that applies to non-relative path. The code will always compile to the same relative pass, independently of the service/computer is being compiled on. 
The gain of using non-relative import is so much higher imo. It's even supported out of the box by NestJs.

Let me know if I misses something, but I haven't seen any large projects using fully relative path, nor a good argument to use them.

**Comment by kamilmysliwiec** [comment]
Yeah just to clarify: it applies only to paths that contain `src` directory (so using `rootDir` + absolute paths in your tsconfig)

**Comment by Val-istar-Guo** [comment]
I had the same problem and couldn't import an esm package. Do I have to change all absolute paths to relative paths.

**Comment by amit78523** [comment]
> I had the same problem and couldn't import an esm package. Do I have to change all absolute paths to relative paths. Is there any solution to solve this problem temporarily.

I am using this command temporarily `nest build && tsc-alias` tsc-alias is a node module.

**Comment by Ttou** [comment]
same problem, https://github.com/microsoft/TypeScript/issues/56350

**Comment by demaisj** [comment]
Encountered this bug today, here's a minimum reproduction repo and steps:
https://github.com/demaisj/nestjs-tsconfig-path-issuerepro
1. `npm install`
2. setup path alias in tsconfig.json (done in repo)
```json
"paths": {
  "@/*": ["./src/*"]
}
```
3. use alias import with .js extension (done in repo)
```ts
import { AppService } from "@/app.service.js";
```
5. `npm run start`
6. observe the import error from nodejs runtime
```
node:internal/modules/cjs/loader:1148
  throw err;
  ^

Error: Cannot find module '@/app.service.js'
```
We can see in the compiled `app.module.js` file that the import alias without js extension is correctly resolved `@/app.controller → ./app.controller` contrary to the import alias with explicit extension`@/app.service.js → @/app.service.js`
Resolving alias with explicit extension would also allow users to work in an ESM environment.

**Comment by JeromeYangtao** [comment]
> > I had the same problem and couldn't import an esm package. Do I have to change all absolute paths to relative paths. Is there any solution to solve this problem temporarily.
> 
> I am using this command temporarily `nest build && tsc-alias` tsc-alias is a node module.

Do you have any solution for nest start

**Comment by kareldonk** [comment]
Is there a way to run tsc-alias as a post build step when doing `nest start`?

**Comment by mkvlrn** [comment]
@kareldonk 
> Is there a way to run tsc-alias as a post build step when doing `nest start`?

Not sure if it is ideal but I think I've arrived at the same point as you (trying to run esm?) and I've settled for this script in my package.json in order to start "dev" mode:

```jsonc
  // ...
  "scripts": {
      // ...
      "dev": "nest build && tsc-alias -p tsconfig.build.json -f && node --watch build/main.js",
      // ...
  }
  // ...
```
